jax/_src/nn/functions.py

```diff
@@ -765,3 +767,178 @@ def hard_silu(x: ArrayLike) -> Array:
+    padded_logits = jnp.where(mask, logits, large_negative_number)
+  else:
+    padded_logits = logits
+
+  if is_causal:
+    T, S = query.shape[-3], key.shape[-3]
+    assert T == S, f'T should equal to S, but got {T} vs {S}'
+    mask = _get_causal_mask(T, logits.dtype)
+    padded_logits = padded_logits + mask
+
+  # Softmax and it is always carried out in fp32.
+  padded_logits = padded_logits.astype(jnp.float32)
+  probs = jax.nn.softmax(padded_logits, axis=-1).astype(key.dtype)
+
+  # Compute the attention context.
+  encoded = jnp.einsum('BNTS,BSNH->BTNH', probs, value)
+  return encoded
+
+def dot_product_attention(
+    query: ArrayLike,
+    key: ArrayLike,
+    value: ArrayLike,
+    bias: ArrayLike | None = None,
+    mask: ArrayLike | None = None,
+    *,
+    scale: float | None = None,
+    is_causal: bool = False,
+    implementation: str | None = None) -> Array | tuple[Array, Array]:
+  r"""Scaled dot product attention function.
+
+  Computes the attention function on Query, Key, and Value tensors:
+
+  .. math ::
+    \mathrm{Attention}(Q, K, V)=\mathrm{softmax}(\frac{QK^T}{\sqrt{d_k}}V)
+
+  If we define :code:`logits` as the output of :math:`QK^T` and the
+  :code:`probs` as the output of :math:`softmax`.
+
+  Throughout this function, we utilize the following uppercase letters to
+  represent the shape of array:
+
+    B = batch size
+    S = length of the key/value (source)
+    T = length of the query (target)
+    N = number of attention heads
+    H = dimensions of each attention head
+
+  Args:
+    query: query array; shape :code:`(BTNH)`
+    key: key array; shape :code:`(BSNH)`
+    value: value array; shape :code:`(BSNH)`
+    bias: optional, bias array to be added to logits; shape broadcastable to
+          :code:`(BNTS)`.
+    mask: optional, mask array used to filter out logits. It is a boolean mask
+          where `True` indicates the element should take part in attention. For
+          an additive mask, users should pass it to `bias`. The shape is
+          broadcastable to :code:`(BNTS)`.
+    scale: scale for the logits. If None, the scale will be set to 1 divided by
+           the square root of query's head dimension (i.e. H).
+    is_causal: If true, causal attention will be applied. Note, some
+               implementations like `xla` will generate a mask tensor and apply
+               it to the logits, but other implementations like `cudnn` will
+               avoid computing the unmasked regions.
+    implementation: A string to control which implementation backend to use.
+                    Supported strings are `xla`, `cudnn` (cuDNN flash
+                    attention). It defaults to `None`, which will automatically
+                    select the best available backend.
+
+  Returns:
+    An array of the attention output with the same shape of :code:`query`.
+
+  Raises:
+    ValueError: if the `implementation` is not a supported option from
```

Comment:
The main raises will be from cuDNN for things that aren't implemented. Are these NotImplemented raises?